src/main.rs

```diff
@@ -5,6 +5,288 @@
+    });
+    abort_handle
+}
+
+#[derive(Debug, From)]
+pub enum INXListenerEvent {
+    Message(Result<inx::proto::Message, Status>),
+    MessagesResult(MessagesResult),
+    Milestone(Result<inx::proto::Milestone, Status>),
+    MilestonesResult(MilestonesResult),
+    Shutdown,
+}
+
+/// A worker that writes messages from [`inx`] to the database.
+pub struct INXListener {
+    db: mongodb::Database,
+    handle: UnboundedSender<INXListenerEvent>,
+    inbox: UnboundedReceiver<INXListenerEvent>,
+}
+
+impl INXListener {
+    fn new(db: mongodb::Database) -> Self {
+        let (handle, inbox) = tokio::sync::mpsc::unbounded_channel();
+        Self { db, handle, inbox }
+    }
+
+    fn handle(&self) -> UnboundedSender<INXListenerEvent> {
+        self.handle.clone()
+    }
+
+    async fn start(mut self) -> Result<(), Error> {
+        info!("INXListener started.");
+        info!("Connecting to INX...");
+        let mut inx_client = InxClient::connect("http://localhost:9029")
+            .await
+            .map_err(|e| Error::other(e.to_string()))?;
+        info!("Connected to INX.");
+        let response = inx_client.read_node_status(NoParams {}).await;
+        info!(
+            "Node status: {:#?}",
+            response.map_err(|e| Error::other(e.to_string()))?.into_inner()
+        );
+
+        let messages_abort = supervise::<INXListenerEvent, MessagesResult, _>(
+            self.handle(),
+            messages(inx_client.clone(), self.handle()),
+        );
+        let milestones_abort = supervise::<INXListenerEvent, MilestonesResult, _>(
+            self.handle(),
+            latest_milestone(inx_client.clone(), self.handle()),
+        );
+        while let Some(evt) = self.inbox.recv().await {
+            match evt {
+                INXListenerEvent::Message(msg) => {
+                    self.handle_message(msg?).await?;
+                }
+                INXListenerEvent::MessagesResult(MessagesResult(res)) => {
+                    if let Err(err) = res {
+                        error!("Error while listening to messages: {}", err);
+                        if self.handle.send(INXListenerEvent::Shutdown).is_err() {
+                            break;
+                        }
+                    }
+                }
+                INXListenerEvent::Milestone(msg) => {
+                    self.handle_milestone(msg?).await?;
+                }
+                INXListenerEvent::MilestonesResult(MilestonesResult(res)) => {
+                    if let Err(err) = res {
+                        error!("Error while listening to milestones: {}", err);
+                        if self.handle.send(INXListenerEvent::Shutdown).is_err() {
+                            break;
+                        }
+                    }
+                }
+                INXListenerEvent::Shutdown => break,
+            }
+        }
+        // This isn't really good enough, due to early exits :(
```

Comment:
We could fix this by creating a wrapper around `AbortHandle` and implementing `Drop` for it so it calls `handle.abort()`. What do you think?